# rvfvtyping: Output

## Introduction

This document describes the output produced by the pipeline. Most of the plots are taken from the MultiQC report, which summarises results at the end of the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

<!-- TODO nf-core: Write this documentation describing your workflow's output -->

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/)
and processes data using the following steps:

* [diamond](#diamond) - Build local diamond database and BLAST query sequences against the database
* [MAFFT](#MAFFT) - Multiple sequence alignment of the query sequences and the reference dataset
* [iqtree](#iqtree) - Phylogenetic tree inference using Maximum Likelihood
* [python](#python) - Filtering input sequence, parsing data, tree format coversion and lineage assignment.
* [R](#R) - Plotting phylogenetic tree, Multiple sequence alignment, SNPs and sequence distance matrix
* [Pipeline information](#pipeline-information) - Report metrics generated during the workflow execution

## diamond

[diamond](https://github.com/bbuchfink/diamond) DIAMOND is a sequence aligner for protein and translated DNA searches, designed for high performance analysis of big sequence data.

For further reading and documentation see the [diamond wiki](https://github.com/bbuchfink/diamond/wiki).

**Output files:**

* `diamond/`
  * `*.dmnd.blastx.txt`: diamond BLAST report containing similarity, taxonomic and metrics
* `db/`
  * `*.dmnd`: diamond formatted database

## MAFFT

[MAFFT](https://mafft.cbrc.jp/alignment/software/index.html) is a multiple sequence alignment program for unix-like operating systems.  It offers a range of multiple alignment methods, L-INS-i (accurate; for alignment of <∼200 sequences), FFT-NS-2 (fast; for alignment of <∼30,000 sequences), etc.


For more information about how to use MAFFT, see [https://mafft.cbrc.jp/alignment/software/manual/manual.html](https://mafft.cbrc.jp/alignment/software/manual/manual.html.

**Output files:**

* `alignments/query/fasta`
  * `*.align.fasta`: alignment file in FASTA format
  * `*.log`: log file of the alignment process

## iqtree

[iqtree](http://www.iqtree.org/) A fast and effective stochastic algorithm to infer phylogenetic trees by maximum likelihood. IQ-TREE compares favorably to RAxML and PhyML in terms of likelihoods with similar computing time

For more information about how to use iqtree, see [http://www.iqtree.org/doc/Tutorial](http://www.iqtree.org/doc/Tutorial.

**Output files:**

* `iqtree/`
  * `*.treefile`: athe ML tree in NEWICK format
  * `*.iqtree`: a textual representation of the final tree
  * `*.contree`: the consensus tree with assigned branch supports where branch lengths are optimized on the original alignment


## Pipeline information

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

**Output files:**

* `pipeline_info/`
  * Reports generated by Nextflow: `execution_report_*`, `execution_timeline_*`, `execution_trace_*` and `pipeline_dag_*`.
  * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.csv`.
  * Documentation for interpretation of results in HTML format: `results_description.html`.
